### PR TITLE
Add notification feature in react client

### PR DESCRIPTION
**Pull request by oozywaters**

### CODE REVIEW

**Comment by mxfactorial** [comment]
may avoid `UserNotFoundException: User does not exist.` e2e teardown error by merging https://github.com/systemaccounting/mxfactorial/commit/883d87f65a74c08187acf331fd771b06510dc250 in `develop`

**Comment by oozywaters** [comment]
Ready for review.

File: react/src/decorators/withNotifcations.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+            // Request n more notifcations
+            () => this.getNotifications(data.cleared.length)
+          )
+        }
+      }
+
+      clearNotifications = async () => {
+        const socket = this.socketRef.current
+        const { notifications } = this.state
+        console.log(socket)
```
Comment (mxfactorial) [inline react/src/decorators/withNotifcations.js]: please remove

**Review by mxfactorial — COMMENTED** [review]
removed in 8d1a7ac7ec6bfd4bf6ede2bc72d0f7974bfd0722

File: react/package.json [hunk 2]
```diff
@@ -13,19 +13,24 @@
     "font-awesome": "^4.7.0",
     "graphql-tag": "^2.10.0",
     "hoist-non-react-statics": "^2.5.5",
+    "jest-websocket-mock": "^2.0.0",
```
Comment (mxfactorial) [inline react/package.json]: `devDependencies` please

File: react/package.json [hunk 3]
```diff
@@ -13,19 +13,24 @@
...
     "react-router-dom": "^4.3.1",
     "react-scripts": "latest",
     "react-svg": "^4.1.8",
+    "react-websocket": "^2.1.0",
     "rxjs": "^6.3.3",
     "styled-components": "^3.3.3",
     "styled-is": "^1.1.3",
-    "uuid": "^3.3.2"
+    "uuid": "^3.3.2",
+    "ws": "^7.2.0"
```
Comment (mxfactorial) [inline react/package.json]: please remove if not used

File: react/package.json [hunk 4]
```diff
@@ -13,19 +13,24 @@
     "font-awesome": "^4.7.0",
     "graphql-tag": "^2.10.0",
     "hoist-non-react-statics": "^2.5.5",
+    "jest-websocket-mock": "^2.0.0",
     "jsonschema": "^1.2.4",
-    "prop-types": "^15.6.1",
+    "mock-socket": "^9.0.2",
```
Comment (mxfactorial) [inline react/package.json]: please remove if not used

**Comment by mxfactorial** [comment]
please add "0 increase in pending notifications from websocket refresh" unit test:

duped-notifications

**Comment by mxfactorial** [comment]
please reset notificationCount on clearAllButton **select** now that all connected clients reset on `clearnotifications` websocket event: https://github.com/systemaccounting/mxfactorial/issues/115#issuecomment-568351290

**Comment by mxfactorial** [comment]
missing e2e test for "notificationCounter decreases by 1" from clearAllButton select in `e2e/private/notificationsMenu/inventory.test.js`. please test notificationCount reset with 2 pending notifications

and please add esimtation to https://github.com/systemaccounting/mxfactorial/issues/115 if underestimated

EDIT:
found [expect(notificationsMenu).toHaveLength(0)](https://github.com/systemaccounting/mxfactorial/blob/0c83fae8756c4964854794255a7c04ec1a7ba85c/react/e2e/private/notificationsMenu/notificationsMenu.test.js#L20). please separate into separate test, and clear 2 pending notifications.
will help contrast between upcoming "decrement on notificationDropDownMenuItem select" case

File: react/src/components/TopNavigation/TopNavigation.module.css [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+.menuItem:last-child {
+    margin-right: 0;
+}
+
+.menuItem_active,
+.menuItem:hover {
+    color: #fff;
+}
+
+.badgeWrapper {
```
Comment (mxfactorial) [inline react/src/components/TopNavigation/TopNavigation.module.css]: dupe. please remove

File: react/package.json (hunk 4 above)
Comment (oozywaters) [reply react/package.json]: This is a peer dependency for [jest-websocket-mock](https://github.com/romgain/jest-websocket-mock#readme).
Moved to `devDependencies` alongside with jest-websocket-mock

File: react/package.json (hunk 3 above)
Comment (oozywaters) [reply react/package.json]: Done

File: react/package.json (hunk 2 above)
Comment (oozywaters) [reply react/package.json]: Done

File: react/src/components/TopNavigation/TopNavigation.module.css (hunk 5 above)
Comment (oozywaters) [reply react/src/components/TopNavigation/TopNavigation.module.css]: Done

**Comment by oozywaters** [comment]
> please add "0 increase in pending notifications from websocket refresh" unit test:

Done

**Comment by oozywaters** [comment]
> please reset notificationCount on clearAllButton **select** now that all connected clients reset on `clearnotifications` websocket event: [#115 (comment)](https://github.com/systemaccounting/mxfactorial/issues/115#issuecomment-568351290)

Done

**Comment by oozywaters** [comment]
> missing e2e test for "notificationCounter decreases by 1" from clearAllButton select in `e2e/private/notificationsMenu/inventory.test.js`. please test notificationCount reset with 2 pending notifications

Done

**Comment by oozywaters** [comment]
Ready for review.

**Comment by mxfactorial** [comment]
thank you for well-crafted code

**Pull request merged**